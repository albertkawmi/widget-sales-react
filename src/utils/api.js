import { DEFAULT_API_URL } from './config'

const BASE_API_URL = process.env.REACT_APP_API_URL || DEFAULT_API_URL

const fetchJson = path =>
  fetch([BASE_API_URL, path].join(''))
    .then(response => {
      if (!response.ok) throw new Error(response.statusText)
      return response.json()
    })

const endpoints = {
  get: {
    clients: () => fetchJson('/clients'),
    sales: () => fetchJson('/sales'),
    fault: () => fetchJson('/fault'),
  }
}

export default {
  get: endpoint => {
    if (!(endpoint in endpoints.get)) {
      return Promise.reject('api.get called with invalid endpoint')
    }
    return endpoints.get[endpoint]()
  }
  // other API methods would go here
}
