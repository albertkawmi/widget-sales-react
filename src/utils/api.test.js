import api from './api'

const mockErrorMsg = 'Something went wrong'
const mockError = new Error(mockErrorMsg)
const mockApiData = ['usually an array']

const mockApiResponse = {
  ok: true,
  json: () => Promise.resolve(mockApiData)
}

const mockApiError = {
  ok: false,
  statusText: mockErrorMsg
}

const clientsRoute = 'https://widget-sales-api.now.sh/clients'
const salesRoute = 'https://widget-sales-api.now.sh/sales'
const faultRoute = 'https://widget-sales-api.now.sh/fault'

const mockFetch = url => {
  switch (url) {
    case clientsRoute:
    case salesRoute:
      return Promise.resolve(mockApiResponse)
    case faultRoute:
    default:
      return Promise.resolve(mockApiError)
  }
}

const fetchSpy = jest.spyOn(global, 'fetch')
  .mockImplementation(mockFetch)

beforeEach(() => {
  jest.clearAllMocks()
})

describe('api.get', () => {
  it('calls fetch on the clients endpoint', () => {
    return api.get('clients')
      .then(() => {
        expect(fetchSpy).toHaveBeenCalledWith(
          'https://widget-sales-api.now.sh/clients'
        )
      })
  })

  it('calls fetch on the sales endpoint', () => {
    return api.get('sales')
      .then(() => {
        expect(fetchSpy).toHaveBeenCalledWith(
          'https://widget-sales-api.now.sh/sales'
        )
      })
  })

  it('calls fetch on the fault endpoint', () => {
    return api.get('fault')
      .catch((error) => {
        expect(fetchSpy).toHaveBeenCalledWith(
          'https://widget-sales-api.now.sh/fault'
        )
        expect(error).toEqual(mockError)
      })
  })

  it('throws for invalid endpoints', () => {
    return api.get('not/an/endpoint')
      .catch((error) => {
        expect(fetchSpy).not.toHaveBeenCalled()
        expect(error).toMatchSnapshot()
      })
  })
})
