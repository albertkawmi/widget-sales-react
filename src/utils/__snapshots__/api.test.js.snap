// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`api.get throws for invalid endpoints 1`] = `"api.get called with invalid endpoint"`;
